File: easy/0661_image_smoother/solution.py
from typing import List


class Solution:
    def imageSmoother(self, img: List[List[int]]) -> List[List[int]]:
        rows = len(img)
        cols = len(img[0])

        offsets = [
            (-1, -1),
            (0, -1),
            (1, -1),
            (-1, 0),
            (0, 0),
            (1, 0),
            (-1, 1),
            (0, 1),
            (1, 1),
        ]

        imgSmoothed = [[0] * cols for _ in range(rows)]

        for row in range(rows):
            for col in range(cols):
                cells = []

                for rowOffset, colOffset in offsets:
                    if (
                        0 <= row + rowOffset < rows
                        and 0 <= col + colOffset < cols
                    ):
                        cells.append(img[row + rowOffset][col + colOffset])

                imgSmoothed[row][col] = sum(cells) // len(cells)

        return imgSmoothed

File: easy/0661_image_smoother/solution.cpp
#include <vector>

class Solution {
public:
  std::vector<std::vector<int>>
  imageSmoother(std::vector<std::vector<int>> &img) {
    const size_t rows = img.size();
    const size_t cols = img[0].size();

    std::vector<std::vector<int>> imgSmoothed(rows, std::vector<int>(cols, 0));

    for (size_t row = 0; row < rows; ++row) {
      for (size_t col = 0; col < cols; ++col) {
        int cellsSum = 0;
        int cellsCount = 0;

        for (size_t r = std::max(0, static_cast<int>(row) - 1);
             r < std::min(rows, row + 2); ++r) {
          for (size_t c = std::max(0, static_cast<int>(col) - 1);
               c < std::min(cols, col + 2); ++c) {
            cellsSum += img[r][c];
            ++cellsCount;
          }
        }

        imgSmoothed[row][col] = cellsSum / cellsCount;
      }
    }

    return imgSmoothed;
  }
};
